### PR TITLE
Add annual reports for accounts

### PR DESCRIPTION
**Pull request by Gargron**
Fixes MAS-213

### CODE REVIEW

**Comment by Ember-ruby** [comment]
> Fixes MAS-213

unless I'm missing something, this doesn't seem to be mentioned anywhere on the [roadmap](https://joinmastodon.org/roadmap) or any currently open issue?

**Comment by renchap** [comment]
We have an internal bug tracker for project-wide features, and this is what MAS-213 references. The public roadmap is generated from it, but has not been updated in a while (I need to look at how to automate it).

File: app/lib/annual_report/percentiles.rb
Comment (ClearlyClaire) [inline app/lib/annual_report/percentiles.rb:1]: I'm not sure I enjoy the idea of encouraging people to compete for followers count.

**Review by ClearlyClaire — COMMENTED** [review]
I have made a few inline comments. In addition, I am concerned about two things:
- I'm concerned with the server load of going through each account's year of posts, this may expensive if your instance has prolific posters
- I'm concerned with the fact this stores user metadata of a somewhat personal nature (such as most used app, most reblogged accounts, interacted with accounts, …) that the user may not know about (because of not using a client that makes use of this API) or may not delete (the current implementation does not provide any way to remove them from the database) save for outright deleting their account

File: app/workers/generate_annual_report_worker.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class GenerateAnnualReportWorker
```
[the remark this answers is not in the document]
Comment (Gargron) [reply app/workers/generate_annual_report_worker.rb:3]: @renchap asked me to remove the scheduling code, so this is for me to call manually from the console.

File: app/lib/annual_report/percentiles.rb
Comment (Gargron) [reply app/lib/annual_report/percentiles.rb:1]: I don't think people will be encouraged to compete based on a metric they see once a year. It's just a fun little piece of trivia.

File: app/workers/generate_annual_report_worker.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (ClearlyClaire) [reply app/workers/generate_annual_report_worker.rb:3]: I see. This lessens the concerns about it being expensive, I guess!

File: app/controllers/api/v1/annual_reports_controller.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    render json: @presenter,
+           serializer: REST::AnnualReportsSerializer,
+           relationships: @relationships
+  end
+
+  def read
+    @annual_report.view!
+    render_empty
+  end
```
[the remark this answers is not in the document]
Comment (jonpincus) [reply app/controllers/api/v1/annual_reports_controller.rb:23]: If there isn't any way to delete, that sounds like a potential problem from a GDPR perspective (and perhaps some American privacy laws like Washington's My Health My Data as well, which applies to sites no matter what size)

**Review by ClearlyClaire — DISMISSED** [review]
I'm still not a fan of the popularity-based information, and I still have concerns regarding the performance cost of generating a large number of these reports, and regarding the fact this essentially holds (small amounts of) user metadata the user can't remove unless they outright delete their account.

However, since these reports are currently not automatically created and can only be manually created through a Rails console, these concerns are not blocker for now, but they probably need to be addressed before those reports get generalized.

**Comment by Gargron** [comment]
Last change adds a percentile calculation based on posts created as per Sam's comment.

**Review by ClearlyClaire — APPROVED** [review]
Repeating my previous review as it is still relevant:
> I'm still not a fan of the popularity-based information, and I still have concerns regarding the performance cost of generating a large number of these reports, and regarding the fact this essentially holds (small amounts of) user metadata the user can't remove unless they outright delete their account.
> 
> However, since these reports are currently not automatically created and can only be manually created through a Rails console, these concerns are not blocker for now, but they probably need to be addressed before those reports get generalized.

**Comment by ShadowJonathan** [comment]
There's absolutely no information as to what this feature is about, and we have no idea what we're merging into our forks at the moment; is this some sort of "mastodon wrapped" feature? And if so, could it be turned off? There is absolutely no reason for this to become a staple feature for a server, it only adds to the cruft and maintenance.

**Comment by renchap** [comment]
This is an experimental feature, to power an eventual yearly recap feature.

This code can not be triggered in any way at the moment, and will be built upon later. Nothing is defined yet, but I can see the need for a server setting to disable it.

**Comment by mdingemanse** [comment]
Marking my surprise that the comments by @ClearlyClaire are not even acknowledged. This is not only an additional server burden, it is also personal and potentially GDPR-relevant info that users will not be aware is being collected and that they should have power to delete or opt out of.

**Comment by renchap** [comment]
As I said above, this is an experimental feature that can not be triggered except through the console. The comments above have not been ignored, but work on the feature has stopped for now. It will evolve in a future PR (or removed).

**Pull request merged**